include/emp/web/Color.hpp

```diff
@@ -0,0 +1,459 @@
+        return strtoll(str.c_str(), nullptr, base);
+    }
+
+    uint8_t parse_css_int(const std::string& str) {  // int or percentage.
+        if (str.length() && str.back() == '%') {
+            return clamp_css_byte(parseFloat(str) / 100.0f * 255.0f);
+        } else {
+            return clamp_css_byte(parseInt(str));
+        }
+    }
+
+    float parse_css_float(const std::string& str) {  // float or percentage.
+        if (str.length() && str.back() == '%') {
+            return clamp_css_float(parseFloat(str) / 100.0f);
+        } else {
+            return clamp_css_float(parseFloat(str));
+        }
+    }
+
+    float css_hue_to_rgb(const float m1, const float m2, float h) {
+        if (h < 0.0f) {
+            h += 1.0f;
+        } else if (h > 1.0f) {
+            h -= 1.0f;
+        }
+
+        if (h * 6.0f < 1.0f) {
+            return m1 + (m2 - m1) * h * 6.0f;
+        }
+        if (h * 2.0f < 1.0f) {
+            return m2;
+        }
+        if (h * 3.0f < 2.0f) {
+            return m1 + (m2 - m1) * (2.0f / 3.0f - h) * 6.0f;
+        }
+        return m1;
+    }
+
+    emp::vector<std::string> split(const std::string& s, char delim) {
+        emp::vector<std::string> elems;
+        std::stringstream ss(s);
+        std::string item;
+        while (std::getline(ss, item, delim)) {
+            elems.push_back(item);
+        }
+        return elems;
+    }
+
+  }
+
+  Color::Color( const std::string& css_str ) {
+   std::string str = css_str;
+
+    // Remove all whitespace, not compliant, but should just be more accepting.
+    emp::remove_whitespace(str);
+
+    // Convert to lowercase.
+    emp::to_lower(str);
+
+    // Ensure sorted and find named color match with lower_bound
+    emp_assert(std::is_sorted(
+        std::begin(color_impl::namedColors), std::end(color_impl::namedColors), 
+        [](const auto& left, const auto& right){
+            return left.name < right.name;
+        }));
+    
+    auto namedColor = std::lower_bound(
+        std::begin(color_impl::namedColors), std::end(color_impl::namedColors), str,
+        [](const auto& left, const auto& right){
+            return left.name < right;
+        });
+
+    if (namedColor->name == str){
+        *this = namedColor->color;
+        (*this).css_str = emp::ColorRGB(r, g, b, a);
+        return;
+    }
+
+    *this = ParseColor(str);
+
+  }
+  
+  // Handle all color detection and parsing from the string
+  Color Color::ParseColor(const std::string& str) {
+      // #abc and #abc123 syntax.
+    if (DetectSyntaxABC(str)) {
+        return ParseABC(str);
+    }
+
+    size_t open_paren = str.find_first_of('(');
+    size_t close_paren = str.find_first_of(')');
+    if (open_paren != std::string::npos && close_paren + 1 == str.length()) {
+        const std::string fname = str.substr(0, open_paren);
+        const std::vector<std::string> color_tokens = color_impl::split(str.substr(open_paren + 1, 
```

Comment:
I would move color_tokens /into/ the ParseRGB and ParseHSL functions instead of computing them here and passing them as params.

Reply:
If we do this, we would have to pass open_paren and close_paren, or compute them both places because we have to know fname to call the right function. Is that still better?